src/Microsoft.AspNetCore.SignalR.Redis/Internal/MsgPackUtil.cs

```diff
@@ -0,0 +1,65 @@
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using MessagePack;
+
+namespace Microsoft.AspNetCore.SignalR.Redis.Internal
+{
+    internal static class MsgPackUtil
+    {
+        public static int ReadArrayHeader(ref ReadOnlyMemory<byte> data)
+        {
+            var arr = GetArray(in data);
+            var val = MessagePackBinary.ReadArrayHeader(arr.Array, arr.Offset, out var readSize);
+            data = data.Slice(readSize);
+            return val;
+        }
+
+        public static int ReadMapHeader(ref ReadOnlyMemory<byte> data)
+        {
+            var arr = GetArray(in data);
+            var val = MessagePackBinary.ReadMapHeader(arr.Array, arr.Offset, out var readSize);
+            data = data.Slice(readSize);
+            return val;
+        }
+
+        public static string ReadString(ref ReadOnlyMemory<byte> data)
+        {
+            var arr = GetArray(in data);
+            var val = MessagePackBinary.ReadString(arr.Array, arr.Offset, out var readSize);
+            data = data.Slice(readSize);
+            return val;
+        }
+
+        public static byte[] ReadBytes(ref ReadOnlyMemory<byte> data)
+        {
+            var arr = GetArray(in data);
+            var val = MessagePackBinary.ReadBytes(arr.Array, arr.Offset, out var readSize);
+            data = data.Slice(readSize);
+            return val;
+        }
+
+        public static int ReadInt32(ref ReadOnlyMemory<byte> data)
+        {
+            var arr = GetArray(in data);
+            var val = MessagePackBinary.ReadInt32(arr.Array, arr.Offset, out var readSize);
+            data = data.Slice(readSize);
+            return val;
+        }
+
+        public static byte ReadByte(ref ReadOnlyMemory<byte> data)
```

Comment:
This cant be efficient

Reply:
I could bypass MessagePack-CSharp completely. It's always going to be a single byte. This is about reading a single-byte integer btw, not about reading large quantities of data.

Reply:
Can you clarify what you mean? The whole pattern or this method in particular? We can require an array, since we know it's always an array, it just means more offset tracking. I can try and benchmark it if you want.

Reply:
We can do that after I guess. Yes, I'm guessing the ref adds more overhead because of the slicing. Doesn't need to be in this PR.

Reply:
Sure. Ideally, we'd get MessagePack-CSharp to support `ReadOnlySpan` natively ;)